register_dwi_dtitk.py

```diff
@@ -0,0 +1,145 @@
+target_mask_path = Path(os.path.abspath(args.target_mask))
+
+output_path = Path(os.path.abspath(args.output_dir))
+
+# Check source files
+if not source_path.exists() or not source_bval_path.exists() or not source_bvec_path.exists():
+  raise FileNotFoundError(f"File {source_path} not found (or missing corresponding bval/bvec)")
+  
+if not source_mask_path.exists():
+  raise FileNotFoundError(f"File {source_mask_path} not found")
+
+# Check target files  
+if not target_path.exists() or not target_bval_path.exists() or not target_bvec_path.exists() :
+  raise FileNotFoundError(f"File {target_path} not found (or missing corresponding bval/bvec)")
+  
+if not target_mask_path.exists():
+  raise FileNotFoundError(f"File {target_mask_path} not found")
+
+# If the output directory doesn't exist let's create it
+if not output_path.exists():
+  os.mkdir(str(output_path))
+  
+# Copy the masks into the output directory so we can modify/resample
+working_source_mask_path = '%s/source_mask.nii.gz' %(output_path)
+subprocess.run(['cp', source_mask_path, working_source_mask_path])
+
+working_target_mask_path = '%s/target_mask.nii.gz' %(output_path)
+subprocess.run(['cp', target_mask_path, working_target_mask_path])
+  
+# === Generate dti volumes  === (this could be made generic so we could plug in different packages/methods)
+
+print("\n  Generating dti volumes...")
+
+source_data, source_affine, source_img = load_nifti(str(source_path), return_img=True)
+source_bvals, source_bvecs = read_bvals_bvecs(str(source_bval_path), str(source_bvec_path))
+
+source_gtab = gradient_table(source_bvals, source_bvecs)
+
+source_mask_data, source_mask_affine, source_mask_img = load_nifti(source_mask_path, return_img=True)
+
+source_tenmodel = dti.TensorModel(source_gtab)
+source_tenfit = source_tenmodel.fit(source_data, mask=source_mask_data)
+
+source_dti_lotri = source_tenfit.lower_triangular()
+
+source_out_dti_filename = '%s/source_dti.nii.gz' %(output_path)
+save_nifti(source_out_dti_filename, source_dti_lotri, source_affine, source_img.header)
+
+target_data, target_affine, target_img = load_nifti(str(target_path), return_img=True)
+target_bvals, target_bvecs = read_bvals_bvecs(str(target_bval_path), str(target_bvec_path))
+
+target_gtab = gradient_table(target_bvals, target_bvecs)
+
+target_mask_data, target_mask_affine, target_mask_img = load_nifti(target_mask_path, return_img=True)
+
+target_tenmodel = dti.TensorModel(target_gtab)
+target_tenfit = target_tenmodel.fit(target_data, mask=target_mask_data)
+
+target_dti_lotri = target_tenfit.lower_triangular()
+
+target_out_dti_filename = '%s/target_dti.nii.gz' %(output_path)
+save_nifti(target_out_dti_filename, target_dti_lotri, target_affine, target_img.header)
+
+# === Preprocess dti for registration with dt-itk ===
+
+print("\n  Preprocessing dti volumes...")
+
+# First we have to scale the diffusivity values to make the units compatible with dti-tk (mean value for CSF around 3)
```

Comment:
Curious how did you figure this out?

What do you mean by mean value for CSF around 3?

Reply:
This info is hidden among the dti-tk documentation: https://dti-tk.sourceforge.net/pmwiki/pmwiki.php?n=Documentation.Diffusivity

I originally assumed it meant the average value of the voxels of CSF around the brain in the DTI (presumably the first component). Looking at their sample data this is the case. But now I think it is referring to the scalar value mean diffusivity. I will calculate that scalar map and take a look.

Reply:
Ohh huh that is easy to miss

> I originally assumed it meant the average value of the voxels of CSF around the brain in the DTI (presumably the first component). Looking at their sample data this is the case. But now I think it is referring to the scalar value mean diffusivity.

Agreed that it is probably referring to mean diffusivity (average of the eigenvalues of the diffusion tensor), though since CSF has isotropic diffusion the first component (I guess it's the xx component) is probably close to the mean diffusivity. 

(a quick way to get mean diffusivity from the tensor is to take the trace and divide by 3)